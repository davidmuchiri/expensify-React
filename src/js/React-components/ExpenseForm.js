import React, { Component } from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';
import 'react-dates/initialize';
import 'react-dates/lib/css/_datepicker.css';

const now = moment();
console.log(now.format('Do MMM, YYYY'));

export default class ExpenseForm extends Component {
	constructor(props) {
		super(props);
		this.state = {
			description: props.expense ? props.expense.description : '',
			note: props.expense ? props.expense.note : '',
			amount: props.expense ? props.expense.amount.toString() : '',
			createdAt: props.expense
				? moment(props.expense.createdAt)
				: moment(),
			calenderFocused: false,
			error: '',
			buttonText: props.expense ? 'Update Expense' : 'Add Expense'
		};
	}

	onDescriptionChange = e => {
		const description = e.target.value;
		this.setState(() => ({ description }));
	};
	onNoteChange = e => {
		const note = e.target.value;
		this.setState(() => ({ note }));
	};
	onAmountChange = e => {
		const amount = e.target.value;
		if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
			this.setState(() => ({ amount }));
		}
	};
	onDateChange = createdAt => {
		if (createdAt) {
			this.setState(() => ({ createdAt }));
		}
	};
	onFocusChange = ({ focused }) => {
		this.setState(() => ({ calenderFocused: focused }));
	};

	onSubmit = e => {
		e.preventDefault();
		if (!this.state.description || !this.state.amount) {
			let error = 'Please provide description and amount';
			this.setState(() => ({ error }));
		} else {
			this.setState(() => ({ error: '' }));
			console.log('submitted');
			this.props.onSubmit({
				description: this.state.description,
				amount: parseFloat(this.state.amount).toFixed(2),
				createdAt: this.state.createdAt.valueOf(),
				note: this.state.note
			});
		}
	};

	render() {
		return (
			<div>
				{this.state.error && <p>{this.state.error}</p>}
				<form onSubmit={this.onSubmit}>
					<input
						type="text"
						placeholder="Description"
						autoFocus
						value={this.state.description}
						onChange={this.onDescriptionChange}
					/>
					<input
						type="text"
						value={this.state.amount}
						placeholder="Amount"
						onChange={this.onAmountChange}
					/>
					<SingleDatePicker
						date={this.state.createdAt}
						onDateChange={this.onDateChange}
						focused={this.state.calenderFocused}
						onFocusChange={this.onFocusChange}
						numberOfMonths={1}
						isOutsideRange={() => false}
					/>
					<textarea
						placeholder="Add a note for expense (optional)"
						value={this.state.note}
						onChange={this.onNoteChange}
					/>
					<button>{this.state.buttonText}</button>
				</form>
			</div>
		);
	}
}
