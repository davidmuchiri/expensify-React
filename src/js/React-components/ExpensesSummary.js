import React from 'react';
import getExpensesTotal from '../redux/selectors/selectExpensesTotal';
import { connect } from 'react-redux';
import selectExpenses from '../redux/selectors/expensesfilter';
import numeral from 'numeral';

export const ExpenseSummary = ({ expensesCount, expensesTotal }) => {
	const expenseWord = expensesCount === 1 ? 'expense' : 'expenses';
	const formattedExpenseTotal = numeral(expensesTotal).format('$0,0.00');

	return (
		<div>
			<h1>
				Viewing {expensesCount} {expenseWord} totalling{' '}
				{formattedExpenseTotal}
			</h1>
		</div>
	);
};

const mapStateToProps = state => {
	const visibleExpenses = selectExpenses(state.expenses, state.filters);
	return {
		expensesCount: visibleExpenses.length,
		expensesTotal: getExpensesTotal(visibleExpenses)
	};
};
export default connect(mapStateToProps)(ExpenseSummary);
