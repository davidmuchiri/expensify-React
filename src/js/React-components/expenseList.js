import React from 'react';
import { connect } from 'react-redux';
import ExpenseListItem from './expenseListItem';
import selectExpenses from '../redux/selectors/expensesfilter';

export const ExpenseList = props => (
	<div>
		{props.expenses.length === 0 ? (
			<p>No expenses</p>
		) : (
			(<h1> Expense List</h1>,
			props.expenses.map(expense => (
				<ExpenseListItem key={expense.id} {...expense} />
			)))
		)}
	</div>
);

const mapStateToProps = state => {
	return {
		expenses: selectExpenses(state.expenses, state.filters)
	};
};

export default connect(mapStateToProps)(ExpenseList);
