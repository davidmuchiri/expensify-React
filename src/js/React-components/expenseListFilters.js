import React, { Component } from 'react';
import { connect } from 'react-redux';
import { DateRangePicker } from 'react-dates';
import 'react-dates/lib/initialize';
import {
	setTextFilter,
	sortByDate,
	sortByAmount,
	setStartDate,
	setEndDate
} from '../redux/actions/filters';

const mapStateToProps = state => {
	return {
		filters: state.filters
	};
};
class ExpenseListFilters extends Component {
	state = {
		calendarFocused: null
	};

	onDatesChange = ({ startDate, endDate }) => {
		this.props.dispatch(setStartDate(startDate));
		this.props.dispatch(setEndDate(endDate));
	};
	onFocusChange = calendarFocused => {
		this.setState(() => ({ calendarFocused }));
	};
	render() {
		return (
			<div>
				<input
					type="text"
					value={this.props.filters.text}
					onChange={e => {
						this.props.dispatch(
							setTextFilter(e.target.value)
						);
					}}
				/>
				<select
					value={this.props.filters.sortBy}
					onChange={e => {
						if (e.target.value === 'date') {
							this.props.dispatch(sortByDate());
						} else if (e.target.value === 'amount') {
							this.props.dispatch(sortByAmount());
						}
					}}>
					<option value="date">Date</option>
					<option value="amount">Amount</option>
				</select>

				<DateRangePicker
					startDate={this.props.filters.startDate}
					endDate={this.props.filters.endDate}
					onDatesChange={this.onDatesChange}
					focusedInput={this.state.calendarFocused}
					onFocusChange={this.onFocusChange}
					numberOfMonths={1}
					isOutsideRange={() => false}
					showClearDates={true}
					startDateId={'start'}
					endDateId={'end'}
				/>
			</div>
		);
	}
}

export default connect(mapStateToProps)(ExpenseListFilters);
