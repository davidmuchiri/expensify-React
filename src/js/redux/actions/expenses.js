import database from '../../../db/firebase';

export const addExpense = expenses => ({ type: 'ADD_EXPENSE', expenses });
export const startAddExpense = (expenseData = {}) => {
	return dispatch => {
		const {
			description = '',
			note = '',
			amount = 0,
			createdAt = 0
		} = expenseData;

		const expense = {
			description,
			note,
			amount,
			createdAt
		};

		return database
			.ref('expenses')
			.push(expense)
			.then(ref => {
				dispatch(addExpense({ id: ref.key, ...expense }));
			})
			.catch(err =>
				console.log('could not add data due to: ', err)
			);
	};
};

export const removeExpense = ({ id } = {}) => ({ type: 'REMOVE_EXPENSE', id });
export const startRemoveExpense = ({ id } = {}) => {
	return dispatch => {
		return database
			.ref(`expenses/${id}`)
			.remove()
			.then(() => dispatch(removeExpense({ id })))
			.catch(err =>
				console.log('could not remove data due to: ', err)
			);
	};
};

//SET_EXPENSES
export const setExpenses = expenses => ({ type: 'SET_EXPENSES', expenses });
export const startSetExpenses = () => {
	return dispatch => {
		return database
			.ref('expenses')
			.once('value')
			.then(snapshot => {
				const expenses = [];
				snapshot.forEach(childSnapshot => {
					expenses.push({
						id: childSnapshot.key,
						...childSnapshot.val()
					});
				});
				dispatch(setExpenses(expenses));
			})
			.catch(err =>
				console.log('could not fetch data due to: ', err)
			);
	};
};

export const editExpense = (id, updates) => ({
	type: 'EDIT_EXPENSE',
	id,
	updates
});
export const startEditExpense = (id, updates) => {
	return dispatch => {
		return database
			.ref(`expenses/${id}`)
			.update(updates)
			.then(() => dispatch(editExpense(id, updates)))
			.catch(err =>
				console.log('could not update data due to: ', err)
			);
	};
};

/**
 * @description how it currently works
 * component calls action generator
 * action generator returns object
 * component dispatches object
 * redux store changes
 */

/**
 * @description how it will work
 * component calls action generator
 * action generator returns function
 * component dispatches function
 * function runs (has the ability to dispatch other actios and do whatever it wants)
 */

/**
 * @description before
 * export const addExpense = ({
	description = '',
	note = '',
	amount = 0,
	createdAt = 0
} = {}) => ({
	type: 'ADD_EXPENSE',
	expenses: {
		id: uuid(),
		description,
		note,
		amount,
		createdAt
	}
});
 */
