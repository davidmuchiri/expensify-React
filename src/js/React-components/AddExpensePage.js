import React from 'react';
import ExpenseForm from './ExpenseForm';
const AddExpensePage = () => (
	<div>
		<h1>Add expense page</h1>
		<ExpenseForm />
	</div>
);

export default AddExpensePage;
