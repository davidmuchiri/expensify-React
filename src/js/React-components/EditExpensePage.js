import React, { Component } from "react";
import { connect } from "react-redux";
import { Link } from "react-router-dom";
import ExpenseForm from "./ExpenseForm";
import FaBack from "react-icons/lib/md/arrow-back";
import {
	startEditExpense,
	startRemoveExpense
} from "../redux/actions/expenses";

export class EditExpensePage extends Component {
	onSubmit = expense => {
		this.props.startEditExpense(this.props.expense.id, expense);
		this.props.history.push("/dashboard");
	};

	onRemove = () => {
		this.props.startRemoveExpense({ id: this.props.expense.id });
		this.props.history.push("/dashboard");
	};

	render() {
		return (
			<div className="container">
				<div className="addExpensepage__panel">
					<Link
						to="/dashboard"
						className="addExpensepage__panel--link">
						<FaBack className="icon" />
					</Link>
					<h5 className="addExpensepage__title">
						Edit expense page
					</h5>
				</div>
				<ExpenseForm
					expense={this.props.expense}
					onSubmit={this.onSubmit}
					onRemove={this.onRemove}
				/>
			</div>
		);
	}
}

const mapDispatchToProps = (dispatch, props) => ({
	startEditExpense: (id, expense) =>
		dispatch(startEditExpense(id, expense)),
	startRemoveExpense: data => dispatch(startRemoveExpense(data))
});

const mapStateToProps = (state, props) => ({
	expense: state.expenses.find(
		expense => expense.id === props.match.params.id
	)
});

export default connect(
	mapStateToProps,
	mapDispatchToProps
)(EditExpensePage);
