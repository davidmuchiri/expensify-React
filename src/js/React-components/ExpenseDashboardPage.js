import React from 'react';
import ExpenseList from './expenseList';
import ExpenseListFilters from './expenseListFilters';
import ExpensesSummary from './ExpensesSummary';

const ExpenseDashboardPage = () => (
	<div>
		<ExpensesSummary />
		<ExpenseListFilters />
		<ExpenseList />
	</div>
);

export default ExpenseDashboardPage;
