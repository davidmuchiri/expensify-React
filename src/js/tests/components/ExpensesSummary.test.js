import React, { Component } from 'react';
import { shallow } from 'enzyme';
import expenses from '../fixtures/testData';
import { ExpenseSummary } from '../../React-components/ExpensesSummary';

test('should render summary component with no expenses', () => {
	const wrapper = shallow(
		<ExpenseSummary expensesCount={1} expensesTotal={235} />
	);
	expect(wrapper).toMatchSnapshot();
});
test('should render summary component with expenses', () => {
	const wrapper = shallow(
		<ExpenseSummary expensesCount={5} expensesTotal={400} />
	);
	expect(wrapper).toMatchSnapshot();
});
