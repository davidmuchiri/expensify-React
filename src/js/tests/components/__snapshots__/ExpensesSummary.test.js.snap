// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render summary component with expenses 1`] = `
<div>
  <h1>
    Viewing 
    5
     
    expenses
     totalling
     
    $400.00
  </h1>
</div>
`;

exports[`should render summary component with no expenses 1`] = `
<div>
  <h1>
    Viewing 
    1
     
    expense
     totalling
     
    $235.00
  </h1>
</div>
`;
